agent/src/main/java/com/microsoft/applicationinsights/agent/internal/agent/AgentImplementation.java

```diff
@@ -136,24 +139,36 @@ private static void appendJarsToBootstrapClassLoader(Instrumentation inst) throw
     }
 
     public static String getAgentJarLocation() throws UnsupportedEncodingException {
+        ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
         try {
-            ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
             if ((systemClassLoader instanceof URLClassLoader)) {
                 for (URL url : ((URLClassLoader)systemClassLoader).getURLs()) {
                     String urlPath = url.getPath();
-                    if (urlPath.indexOf(AGENT_JAR_PREFIX) != -1) {
-                        int index = urlPath.lastIndexOf('/');
-                        urlPath = urlPath.substring(0, index + 1);
-                        return urlPath;
+                    if (urlPath.charAt(0) == '/') {
+                        urlPath = urlPath.substring(1);
+                    }
+                    Path path = Paths.get(urlPath);
+                    if (path.getFileName().toString().startsWith(AGENT_JAR_PREFIX)) {
+                        return path.getParent().toString();
                     }
                 }
             }
         } catch (Throwable throwable) {
             InternalAgentLogger.INSTANCE.logAlways(InternalAgentLogger.LoggingLevel.ERROR, "Error while trying to fetch Jar Location, Exception: " + throwable.getMessage());
         }
 
-        String path = AgentImplementation.class.getProtectionDomain().getCodeSource().getLocation().getPath();
-        return URLDecoder.decode(path, "UTF-8");
+        String stringPath = AgentImplementation.class.getProtectionDomain().getCodeSource().getLocation().getPath();
+        if (stringPath.charAt(0) == '/') {
+            stringPath = stringPath.substring(1);
+        }
+        Path path = Paths.get(stringPath);
```

Comment:
The `java.nio.*` classes were introduced in 1.7. It's my understanding we still want compatibility with 1.6. If that's still the case, this needs to be reworked. @grlima, can you confirm?